transaction/std/src/signed_contingent_input_builder.rs

```diff
@@ -0,0 +1,2272 @@
+    input_credentials: InputCredentials,
+    /// Global indices for the tx out's in the ring
+    tx_out_global_indices: Vec<u64>,
+    /// The outputs created by the transaction, and associated secrets
+    outputs_and_secrets: Vec<(TxOut, OutputSecret)>,
+    /// The tombstone_block value, a block index in which the transaction
+    /// expires, and can no longer be added to the blockchain
+    tombstone_block: u64,
+    /// The source of validated fog pubkeys used for this transaction
+    fog_resolver: FPR,
+    /// The limit on the tombstone block value imposed pubkey_expiry values in
+    /// fog pubkeys used so far
+    fog_tombstone_block_limit: u64,
+    /// An policy object implementing MemoBuilder which constructs memos for
+    /// this transaction.
+    ///
+    /// This is an Option in order to allow working around the borrow checker.
+    /// Box<dyn ...> is used because having more generic parameters creates more
+    /// types that SDKs must bind to if they support multiple memo builder
+    /// types.
+    memo_builder: Option<Box<dyn MemoBuilder + 'static + Send + Sync>>,
+}
+
+impl<FPR: FogPubkeyResolver> SignedContingentInputBuilder<FPR> {
+    /// Initializes a new SignedContingentInputBuilder.
+    ///
+    /// # Arguments
+    /// * `block_version` - The block version rules to use when signing the
+    ///   input
+    /// * `input_credentials` - Credentials for the input we are signing
+    /// * `tx_out_global_indices` - Global indices for the tx out's in the ring
+    /// * `fog_resolver` - Source of validated fog keys to use with this
+    ///   transaction
+    /// * `memo_builder` - An object which creates memos for the TxOuts in this
+    ///   transaction
+    pub fn new<MB: MemoBuilder + 'static + Send + Sync>(
+        block_version: BlockVersion,
+        input_credentials: InputCredentials,
+        tx_out_global_indices: Vec<u64>,
+        fog_resolver: FPR,
+        memo_builder: MB,
+    ) -> Self {
+        Self::new_with_box(
+            block_version,
+            input_credentials,
+            tx_out_global_indices,
+            fog_resolver,
+            Box::new(memo_builder),
+        )
+    }
+
+    /// Initializes a new SignedContingentInputBuilder, using a Box<dyn
+    /// MemoBuilder> instead of statically typed
+    ///
+    /// # Arguments
+    /// * `block_version` - The block version to use when signing the input
+    /// * `input_credentials` - Credentials for the input we are signing
+    /// * `tx_out_global_indices` - Global indices for the tx out's in the ring
+    /// * `fog_resolver` - Source of validated fog keys to use with this signed
+    ///   contingent input
+    /// * `memo_builder` - An object which creates memos for the TxOuts in this
+    ///   signed contingent input
+    pub fn new_with_box(
+        block_version: BlockVersion,
+        input_credentials: InputCredentials,
+        tx_out_global_indices: Vec<u64>,
```

Comment:
Might be worth validating that `tx_out_global_indices.len() == input_credentials.ring.len()`.